/*
 * messages.proto - Message type declarations.
 * Copyright (C) 2019-2020 University of Texas
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

syntax = "proto3";
option java_multiple_files = false;
option java_package = "org.astria.rpc";
option java_outer_classname = "Messages";

message IntegerArray
{
    repeated int32 array = 1 [packed = true];
}

message DoubleArray
{
    repeated double array = 1 [packed = true];
}

message Parameter
{
    string name = 1;
    double value = 2;
    double min = 3;
    double max = 4;
    string estimation = 5;
}

message Facet
{
    repeated double normal = 1 [packed = true];
    double area = 2;
}

message Maneuver
{
    string time = 1;
    string trigger_event = 2;
    repeated double trigger_params = 3 [packed = true];
    string maneuver_type = 4;
    repeated double maneuver_params = 5 [packed = true];
}

message Station
{
    double latitude = 1;
    double longitude = 2;
    double altitude = 3;
    double azimuth_bias = 4;
    double elevation_bias = 5;
    double range_bias = 6;
    double range_rate_bias = 7;
    double right_ascension_bias = 8;
    double declination_bias = 9;
    repeated double position_bias = 10 [packed = true];
    repeated double position_velocity_bias = 11 [packed = true];
    string bias_estimation = 12;
}

message MeasurementSetting
{
    bool two_way = 1;
    repeated double error = 2 [packed = true];
}

message Settings
{
    double rso_mass = 1;
    double rso_area = 2;
    repeated Facet rso_facets = 3;
    repeated double rso_solar_array_axis = 4 [packed = true];
    double rso_solar_array_area = 5;
    string rso_attitude_provider = 6;
    repeated double rso_spin_velocity = 7 [packed = true];
    repeated double rso_spin_acceleration = 8 [packed = true];

    int32 gravity_degree = 9;
    int32 gravity_order = 10;
    int32 ocean_tides_degree = 11;
    int32 ocean_tides_order = 12;
    bool third_body_sun = 13;
    bool third_body_moon = 14;
    bool solid_tides_sun = 15;
    bool solid_tides_moon = 16;

    string drag_model = 17;
    Parameter drag_coefficient = 18;
    repeated IntegerArray drag_MSISE_flags = 19;
    double drag_exp_rho0 = 20;
    double drag_exp_H0 = 21;
    double drag_exp_Hscale = 22;

    bool rp_sun = 23;
    Parameter rp_coeff_reflection = 24;
    double rp_coeff_absorption = 25;

    repeated Maneuver maneuvers = 26;

    string prop_start = 27;
    string prop_end = 28;
    double prop_step = 29;
    repeated double prop_initial_state = 30 [packed = true];
    repeated string prop_initial_TLE = 31;
    string prop_inertial_frame = 32;
    string prop_step_handler_start_time = 33;
    string prop_step_handler_end_time = 34;

    double integ_min_time_step = 35;
    double integ_max_time_step = 36;
    double integ_abs_tolerance = 37;
    double integ_rel_tolerance = 38;

    bool sim_measurements = 39;
    bool sim_skip_unobservable = 40;
    bool sim_include_extras = 41;
    bool sim_include_station_state = 42;

    map<string, Station> stations = 43;
    map<string, MeasurementSetting> measurements = 44;

    string estm_filter = 45;
    repeated double estm_covariance = 46 [packed = true];
    repeated double estm_process_noise = 47 [packed = true];
    double estm_DMC_corr_time = 48;
    double estm_DMC_sigma_pert = 49;
    Parameter estm_DMC_acceleration = 50;
    double estm_outlier_sigma = 51;
    int32 estm_outlier_warmup = 52;
}

message SettingsArray
{
    repeated Settings array = 1;
}

message Measurement
{
    string time = 1;
    string station = 2;

    double azimuth = 3;
    double elevation = 4;

    double range = 5;
    double range_rate = 6;

    double right_ascension = 7;
    double declination = 8;

    repeated double position = 9 [packed = true];
    repeated double position_velocity = 10 [packed = true];

    double atmospheric_density = 11;
    repeated double acceleration_gravity = 12 [packed = true];
    repeated double acceleration_drag = 13 [packed = true];
    repeated double acceleration_ocean_tides = 14 [packed = true];
    repeated double acceleration_solid_tides = 15 [packed = true];
    repeated double acceleration_third_bodies = 16 [packed = true];
    repeated double acceleration_radiation_pressure = 17 [packed = true];
    repeated double acceleration_thrust = 18 [packed = true];
    repeated double station_state = 19 [packed = true];

    repeated double true_state_cartesian = 20 [packed = true];
    double true_state_sma = 21;
    double true_state_ecc = 22;
    double true_state_inc = 23;
    double true_state_raan = 24;
    double true_state_argp = 25;
    double true_state_mean_anom = 26;
    double true_state_ex = 27;
    double true_state_ey = 28;
    double true_state_hx = 29;
    double true_state_hy = 30;
    double true_state_lm = 31;
}

message MeasurementArray
{
    repeated Measurement array = 1;
}

message Measurement2DArray
{
    repeated MeasurementArray array = 1;
}

message InterpolateEphemerisInput
{
    string source_frame = 1;
    repeated string time = 2;
    repeated DoubleArray ephem = 3;
    int32 num_points = 4;
    string dest_frame = 5;
    string interp_start = 6;
    string interp_end = 7;
    double step_size = 8;
}

message InterpolationOutput
{
    string time = 1;
    repeated double state = 2 [packed = true];
}

message InterpolationOutputArray
{
    repeated InterpolationOutput array = 1;
}

message PropagationOutput
{
    string time = 1;
    repeated DoubleArray states = 2;
}

message PropagationOutputArray
{
    repeated PropagationOutput array = 1;
}

message DetermineOrbitInput
{
    Settings config = 1;
    repeated Measurement measurements = 2;
}

message EstimationOutput
{
    string time = 1;
    string station = 2;
    repeated double estimated_state  = 3 [packed = true];
    repeated DoubleArray propagated_covariance = 4;
    repeated DoubleArray innovation_covariance = 5;
    repeated DoubleArray estimated_covariance = 6;
    map<string, DoubleArray> pre_fit = 7;
    map<string, DoubleArray> post_fit = 8;
}

message EstimationOutputArray
{
    repeated EstimationOutput array = 1;
}

message TransformFrameInput
{
    string src_frame = 1;
    string time = 2;
    repeated double pva = 3 [packed = true];
    string dest_frame = 4;
}

message ImportTDMInput
{
    string file_name = 1;
    string file_format = 2;
}

message AnglesInput
{
    string frame = 1;
    repeated string time = 2;
    double latitude = 3;
    double longitude = 4;
    double altitude = 5;
    repeated double angle1 = 6 [packed = true];
    repeated double angle2 = 7 [packed = true];
}
